src/dvc_data/build.py

```diff
@@ -114,33 +179,21 @@ def _build_tree(
         unit="obj",
         desc=f"Building data objects from {relpath}",
         disable=no_progress_bar,
+        total=-1,
     ) as pbar:
-        for root, _, fnames in walk_iter:
-            if DefaultIgnoreFile in fnames:
-                raise IgnoreInCollectedDirError(
-                    DefaultIgnoreFile, fs.path.join(root, DefaultIgnoreFile)
-                )
-
-            # NOTE: we know for sure that root starts with path, so we can use
-            # faster string manipulation instead of a more robust relparts()
-            rel_key: Tuple[Optional[Any], ...] = ()
-            if root != path:
-                rel_key = tuple(root[len(path) + 1 :].split(fs.sep))
-
-            for fname in fnames:
-                if fname == "":
-                    # NOTE: might happen with s3/gs/azure/etc, where empty
-                    # objects like `dir/` might be used to create an empty dir
-                    continue
-
-                pbar.update()
-                meta, obj = _build_file(
-                    f"{root}{fs.sep}{fname}", fs, name, odb=odb, **kwargs
-                )
-                key = (*rel_key, fname)
-                tree.add(key, meta, obj.hash_info)
-                tree_meta.size += meta.size or 0
-                tree_meta.nfiles += 1
+        new, unchanged_entries = _diff_state_to_dir(
```

Comment:
btw, the fact that we have a hash for a file doesn't mean that we have an object built for it in odb already, so we will still need to call `_build_file` even for unchanged files, but we could pass hash_info to it instead of making it call up state again.

Reply:
yeah, I noticed that but I am not sure how to fix this (considering we may also need to use `_upload_file()`).

`odb.add()` also drops the performance from 75k obj/s to 60k obj/s. :(

Reply:
@skshetry So what do you suggest?